kotlin/find-connected-features-in-utility-networks/src/main/java/com/esri/arcgisruntime/sample/findconnectedfeaturesinutilitynetworks/MainActivity.kt

```diff
@@ -0,0 +1,458 @@
+      .filter { it.code == identifiedFeature.attributes["assettype"].toString().toInt() }[0]
+      .let { utilityAssetType ->
+        // get the list of terminals for the feature
+        val terminals = utilityAssetType.terminalConfiguration.terminals
+        // if there is only one terminal, use it to create a utility element
+        when (terminals.size) {
+          1 -> {
+            utilityNetwork.createElement(identifiedFeature, terminals[0])
+          }
+          // if there is more than one terminal, prompt the user to select one
+          else -> {
+            // get a list of terminal names from the terminals
+            val terminalNames = utilityAssetType.terminalConfiguration.terminals.map { it.name }
+
+            AlertDialog.Builder(this).apply {
+              setTitle("Select utility terminal:")
+              setItems(terminalNames.toTypedArray()) { _, which ->
+                val utilityElement =
+                  utilityNetwork.createElement(identifiedFeature, terminals[which])
+
+                // add the utility element to the map
+                addUtilityElementToMap(
+                  identifiedFeature,
+                  identifiedFeature.geometry as? Point,
+                  utilityElement
+                )
+
+                // show the utility element name in the UI
+                showTerminalNameInStatusLabel(utilityElement.terminal)
+              }
+
+            }.show()
+          }
+        }
+      }
+  }
+
+  /**
+   * Shows the name of a UtilityTerminal in the status label in the UI.
+   *
+   * @param terminal
+   */
+  private fun showTerminalNameInStatusLabel(terminal: UtilityTerminal) {
+    statusTextView.text =
+      getString(R.string.terminal_name, if (terminal.name != "") terminal.name else "default")
+  }
+
+  /**
+   * Uses the elements selected as starting locations and (optionally) barriers to perform a connected trace,
+   * then selects all connected elements found in the trace to highlight them.
+   */
+  fun traceUtilityNetwork(view: View) {
+    // check that the utility trace parameters are valid
+    if (startingLocations.isEmpty()) {
+      reportError("No starting locations provided for trace.")
+      return
+    }
+
+    // show the progress indicator and update the status text
+    progressIndicator.visibility = View.VISIBLE
+    statusTextView.text = getString(R.string.find_connected_features_message)
+    disableButtons()
+
+    // create utility trace parameters for the given trace type
+    val traceType = UtilityTraceType.valueOf(traceTypeSpinner.selectedItem.toString())
+    with(UtilityTraceParameters(traceType, startingLocations)) {
+      // if any barriers have been created, add them to the parameters
+      this.barriers.addAll(barriers)
+      // set the trace configuration using the tier from the utility domain network
+      this.traceConfiguration = mediumVoltageTier?.traceConfiguration
```

Comment:
I would prefer to set up the UtilityTraceParameters object using `apply`, then subsequently call `traceAsync` with it. But again, might just be personal preference. It's hard to come up with a consistent guideline for the use of scope functions...